design/fips-compliance.md

```diff
@@ -0,0 +1,199 @@
+ 
+Using the RHEL go toolchain has its infrastructure complexity as we will need
+ to build a new pipeline for building the project on a subscribed RHEL machine.
+ 
+Luckily, Google maintains the BoringSSL which is very easy to use and gives
+ us FIPS compliancy with minimal work. All that needs to be done is:
+ - Replace the base image from `golang` to the corresponding `goboring/golang
+ `. For example, `golang:1.12` should turn into `goboring/golang:1.12.17b4`
+ - Enable cgo by setting `CGO_ENABLED=1` in the env
+ - Verify that the `authenticator` binary uses the BoringCrypto libraries by
+  running `go tool nm authenticator` and verifying that it has symbols named
+   *_Cfunc__goboringcrypto_*
+   
+More info on the required changes can be found [here](https://github.com/cyberark/conjur-authn-k8s-client/pull/97)
+
+### Non-FIPS compliant image
+
+We should decide if we still want to release a version of the `conjur-authn-k8s-client`
+that is not FIPS compliant (i.e that uses `golang` as the base image). Let's
+look into some pros and cons of both approaches.
+ 
+ | **Option**                             | **Implications**                                                                                                                                                                                                                                                                                                                                                        |
+ |----------------------------------------|-------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
+ | Release only `goboring/golang` version | - Customers use the client and don't require FIPS compliancy, will now miss libraries that are available in `golang` but are not available in `goboring/golang`                                                                                                                                                                                                       |
+ | Release both versions                  | <ul><li>One more Dockerfile to maintain. Any change made to the original Dockerfile should be done also to `Dockerfile.fips` which can be easily forgotten</li><li>Another version to test. Our UTs will stay the same but we will need to run integration tests on both images, in order to release them. This adds complexity to our tests and doubles the tests run-time</li></ul> |
+
+From the table above, I think that we should release only the FIPS compliant
+version. The advantages that customers will get by using the `golang` image
+as the base image is not worth the maintenance burden that Releasing both
+versions will bring to the development team.
```

Comment:
Will it be possible to move from FIPS to non-FIPS with a flag?
It is not mandatory for now, but we should understand the options for later stage if needed.

Reply:
IMO that isn't a rabbit hole we want to go down... that would mean we will need to support two images and these two images could definitely diverge.  If we need functionality offered by a library that is not available to `goboring/golang` then we need to do double the work to find overlap for both base images

Reply:
@hilagross that will not be possible as the FIPS/non-FIPS decision is in the base image that builds the authenticator image. Once the image is built then we are already FIPS/non-FIPS compliant.

Reply:
@orenbm - So from what I'm understanding, if the performance requires to have FIPS and non-FIPS the user will have to choose in advance the right image.

@Tovli - I have no problem with goboring image, but if the performance with FIPS is degrading we should make is non-FIPS and test the performance than as well, if the degradation caused by goboring we should not use it for non-FIPS.

Reply:
@hilagross so from your point of view, the question whether we should have both images relies on the performance test?

degradation is more than 5% -> release 2 images
degradation is less than 5% -> release only FIPS image